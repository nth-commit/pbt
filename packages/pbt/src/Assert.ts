import { Property } from 'pbt-properties';
import { run, RunConfig, RunResult } from './Run';

export type AssertionJournal = string[];

export const assert = <Values extends any[]>(p: Property<Values>, config?: Partial<RunConfig>): AssertionJournal => {
  const result = run(p, config);
  switch (result.kind) {
    case 'success':
      return [];
    case 'failure':
      return [...renderTitle(result), ...renderReproduction(result), ...renderCounterexample(result)];
    /* istanbul ignore next */
    default:
      throw new Error(`Unhandled result: ${JSON.stringify(result)}`);
  }
};

const renderTitle = (result: RunResult.Failure): string[] => [
  `Property failed after ${result.iterationsCompleted} test(s)`,
];

const renderReproduction = (result: RunResult.Failure): string[] => {
  let str = `Reproduction: { "seed": ${result.seed}, "size": ${result.size}`;

  if (result.counterexample.shrinkPath) {
    str += `, "shrinkPath": "${result.counterexample.shrinkPath}"`;
  }

  str += ` }`;

  return [str];
};

const renderCounterexample = (result: RunResult.Failure): string[] => [
  `Counterexample: ${JSON.stringify(result.counterexample.values)}`,
];
