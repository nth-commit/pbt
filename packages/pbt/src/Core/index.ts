export * from './Seed';
export * from './Size';
export * from './Tree';
export * from './RandomStream';
