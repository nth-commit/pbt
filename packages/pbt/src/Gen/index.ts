export * from './Imports';
export * from './Shrink';
export * from './Gen';
export * from './Operators';
export * from './Number';
export * from './Collections';
