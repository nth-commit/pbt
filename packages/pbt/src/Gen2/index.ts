export * from './Imports';
export * from './Shrink';
export * from './Gen';
export { GenIteration } from './GenFunction';
export * from './Range';
