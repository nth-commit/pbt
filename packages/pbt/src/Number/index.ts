export * from './Calculator';
export * from './BigJsCalculator';
export * from './NativeCalculator';
