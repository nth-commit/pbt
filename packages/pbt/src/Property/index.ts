export * from './Imports';
export * from './PropertyIteration';
export * from './PropertyFunction';
export * from './reproduceFailure';
export * from './Property';
export * from './PropertyResult';
