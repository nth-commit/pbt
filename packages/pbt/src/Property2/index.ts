/* istanbul ignore file */

export { Property, PropertyFunction, ShrinkIteration, Counterexample, PropertyIteration } from './Abstractions';
export { property } from './Property';
