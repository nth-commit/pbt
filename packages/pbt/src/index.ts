export { RandomStream, Seed, Size } from './Core';
export { GenTree, GenTreeNode, Complexity, CalculateComplexity } from './GenTree';
export { Gen, Shrink, Shrinker, GenIteration } from './Gen';
export { property, Property, PropertyIteration, PropertyFunction, ShrinkIteration, Counterexample } from './Property';
export {
  check,
  CheckConfig,
  CheckResult,
  sample,
  sampleTrees,
  SampleConfig,
  Sample,
  assert,
  AssertConfig,
  PbtAssertionError,
} from './Runners';
