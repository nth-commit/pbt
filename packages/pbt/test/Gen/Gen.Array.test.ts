import fc from 'fast-check';
import * as dev from '../../src';
import * as domainGen from '../Helpers/domainGen';

const genArrayLength = () => domainGen.integer({ min: 0, max: 10 });

test('snapshot', () => {
  for (let i = 0; i <= 10; i++) {
    const seed = 0;
    const gen = dev.Gen.array(dev.Gen.integer().between(0, 10));

    const sample = dev.sampleTrees(gen, { seed, size: i * 10, iterations: 1 });

    expect(dev.GenTree.format(sample.values[0])).toMatchSnapshot(i.toString());
  }
});

test('Gen.array(gen).ofMaxLength(x).growBy(s) *produces* arrays with length equal to *oracle* Gen.integer().between(0, x).growBy(s)', () => {
  fc.assert(
    fc.property(
      domainGen.sampleConfig(),
      domainGen.gen(),
      genArrayLength(),
      domainGen.scaleMode(),
      (config, elementGen, x, s) => {
        const genArray = dev.Gen.array(elementGen)
          .ofMaxLength(x)
          .growBy(s)
          .map((arr) => arr.length);
        const genInteger = dev.Gen.integer().between(0, x).growBy(s);

        expect(dev.sample(genArray, { ...config, iterations: 1 })).toEqual(
          dev.sample(genInteger, { ...config, iterations: 1 }),
        );
      },
    ),
  );
});

describe('defaults', () => {
  test('default(min) = 0', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), (config, elementGen) => {
        const genDefault = dev.Gen.array(elementGen);
        const genAlt = dev.Gen.array(elementGen).ofMinLength(0);

        expect(dev.sample(genDefault, config)).toEqual(dev.sample(genAlt, config));
      }),
    );
  });

  test('default(max) = 10', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), (config, elementGen) => {
        const genDefault = dev.Gen.array(elementGen);
        const genAlt = dev.Gen.array(elementGen).ofMaxLength(10);

        expect(dev.sample(genDefault, config)).toEqual(dev.sample(genAlt, config));
      }),
    );
  });

  test('default(scale) = linear', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), (config, elementGen) => {
        const genDefault = dev.Gen.array(elementGen);
        const genAlt = dev.Gen.array(elementGen).growBy('linear');

        expect(dev.sample(genDefault, config)).toEqual(dev.sample(genAlt, config));
      }),
    );
  });
});

describe('equivalent APIs', () => {
  test('Gen.array(gen) = gen.array()', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), (config, elementGen) => {
        const genArray = dev.Gen.array(elementGen);
        const genArrayAlt = elementGen.array();

        expect(dev.sample(genArray, config)).toEqual(dev.sample(genArrayAlt, config));
      }),
    );
  });

  test('Gen.array(gen).betweenLengths(x, y) = Gen.array(gen).betweenLengths(y, x)', () => {
    fc.assert(
      fc.property(
        domainGen.sampleConfig(),
        domainGen.gen(),
        genArrayLength(),
        genArrayLength(),
        (config, elementGen, x, y) => {
          const genArray = dev.Gen.array(elementGen).betweenLengths(x, y);
          const genArrayAlt = dev.Gen.array(elementGen).betweenLengths(y, x);

          expect(dev.sample(genArray, config)).toEqual(dev.sample(genArrayAlt, config));
        },
      ),
    );
  });

  test('Gen.array(gen).ofMinLength(x).ofMaxLength(y) = Gen.array(gen).betweenLengths(x, y)', () => {
    fc.assert(
      fc.property(
        domainGen.sampleConfig(),
        domainGen.gen(),
        genArrayLength(),
        genArrayLength(),
        (config, elementGen, x, y) => {
          const genArray = dev.Gen.array(elementGen).ofMinLength(x).ofMaxLength(y);
          const genArrayAlt = dev.Gen.array(elementGen).betweenLengths(x, y);

          expect(dev.sample(genArray, config)).toEqual(dev.sample(genArrayAlt, config));
        },
      ),
    );
  });

  test('Gen.array(gen).ofLength(x) = Gen.array(gen).betweenLengths(x, x)', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), genArrayLength(), (config, elementGen, x) => {
        const genArray = dev.Gen.array(elementGen).ofLength(x);
        const genArrayAlt = dev.Gen.array(elementGen).betweenLengths(x, x);

        expect(dev.sample(genArray, config)).toEqual(dev.sample(genArrayAlt, config));
      }),
    );
  });
});

describe('errors', () => {
  test('Gen.array(gen).ofMinLength(x), x ∉ ℤ *produces* error; minimum must be an integer', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), domainGen.decimalWithAtLeastOneDp(), (config, gen, x) => {
        const genArray = dev.Gen.array(gen).ofMinLength(x);

        expect(() => dev.sample(genArray, config)).toThrowError('Minimum must be an integer');
      }),
    );
  });

  test('Gen.array(gen).ofMaxLength(x), x ∉ ℤ *produces* error; maximum must be an integer', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), domainGen.decimalWithAtLeastOneDp(), (config, gen, x) => {
        const genArray = dev.Gen.array(gen).ofMaxLength(x);

        expect(() => dev.sample(genArray, config)).toThrow('Maximum must be an integer');
      }),
    );
  });

  test('Gen.array(gen).ofMinLength(x), x < 0 *produces* error; minimum must be at least 0', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), domainGen.integer({ max: -1 }), (config, gen, x) => {
        const genArray = dev.Gen.array(gen).ofMinLength(x);

        expect(() => dev.sample(genArray, config)).toThrow('Minimum must be at least 0');
      }),
    );
  });

  test('Gen.array(gen).ofMaxLength(x), x < 0 *produces* error; maximum must be at least 0', () => {
    fc.assert(
      fc.property(domainGen.sampleConfig(), domainGen.gen(), domainGen.integer({ max: -1 }), (config, gen, x) => {
        const genArray = dev.Gen.array(gen).ofMaxLength(x);

        expect(() => dev.sample(genArray, config)).toThrow('Maximum must be at least 0');
      }),
    );
  });
});

describe('shrinks', () => {
  test('Gen.array(gen).ofMinLength(x) *produces* arrays that shrink to length = x', () => {
    fc.assert(
      fc.property(domainGen.minimalConfig(), domainGen.gen(), genArrayLength(), (config, gen, x) => {
        const genArray = dev.Gen.array(gen).ofMinLength(x);

        const min = dev.minimal(genArray, config);

        expect(min).toHaveLength(x);
      }),
    );
  });
});
