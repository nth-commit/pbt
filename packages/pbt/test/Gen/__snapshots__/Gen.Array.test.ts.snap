// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: 0 1`] = `" (c = 0)"`;

exports[`snapshot: 1 1`] = `" (c = 0)"`;

exports[`snapshot: 2 1`] = `
"1,2 (c = 50)
.. (c = 0)
..1 (c = 20)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
..2 (c = 30)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
..0,2 (c = 40)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
....2 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
..1,0 (c = 30)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
.... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
....0 (c = 10)
...... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
..1,1 (c = 40)
.... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)"
`;

exports[`snapshot: 3 1`] = `" (c = 0)"`;

exports[`snapshot: 4 1`] = `
"1,3,0,1 (c = 90)
..0,1,1,3 (c = 90)
.... (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,1,1 (c = 50)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0,1 (c = 30)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........0,0 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........0 (c = 10)
............ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,1 (c = 30)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........0,0 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........0 (c = 10)
............ (c = 0)
......1,1 (c = 40)
........ (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........0,1 (c = 30)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........0,0 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............0 (c = 10)
.............. (c = 0)
........1,0 (c = 30)
..........0,1 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
............0,0 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............0 (c = 10)
................ (c = 0)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)"
`;

exports[`snapshot: 5 1`] = `
"1,2 (c = 50)
.. (c = 0)
..1 (c = 20)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
..2 (c = 30)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
..0,2 (c = 40)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
....2 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
..1,0 (c = 30)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
.... (c = 0)
....0 (c = 10)
...... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
.... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
....0 (c = 10)
...... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
..1,1 (c = 40)
.... (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
....1 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)"
`;

exports[`snapshot: 6 1`] = `
"0,1,0,0 (c = 50)
..0,0,0,1 (c = 50)
.... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
....0,0,0 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,0,1 (c = 40)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
......0 (c = 10)"
`;

exports[`snapshot: 7 1`] = `
"7,0,0,4 (c = 150)
..0,0,4,7 (c = 150)
.... (c = 0)
....0,0 (c = 20)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0 (c = 10)
........ (c = 0)
....0,0,4 (c = 70)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
......0 (c = 10)
........ (c = 0)
......4 (c = 50)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........2 (c = 30)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
........3 (c = 40)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
......0,4 (c = 60)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........4 (c = 50)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..........3 (c = 40)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
............2 (c = 30)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)
................ (c = 0)
................0 (c = 10)
.................. (c = 0)
........0,0 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........0 (c = 10)
............ (c = 0)
........0,2 (c = 40)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..........0,0 (c = 20)"
`;

exports[`snapshot: 8 1`] = `
"7,8,4,6,8,3,8,6 (c = 580)
..3,4,6,6,7,8,8,8 (c = 580)
.... (c = 0)
....3,4,6,6 (c = 230)
...... (c = 0)
......3,4 (c = 90)
........ (c = 0)
........3 (c = 40)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
........4 (c = 50)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..........3 (c = 40)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
............2 (c = 30)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)
................ (c = 0)
................0 (c = 10)
.................. (c = 0)
........0,4 (c = 60)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........4 (c = 50)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............2 (c = 30)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)
................ (c = 0)
................0 (c = 10)
.................. (c = 0)
............3 (c = 40)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)
................ (c = 0)
................0 (c = 10)
.................. (c = 0)
..............2 (c = 30)
................ (c = 0)
................0 (c = 10)
.................. (c = 0)
................1 (c = 20)
.................. (c = 0)
..................0 (c = 10)
.................... (c = 0)
..........0,0 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............0 (c = 10)
.............. (c = 0)
..........0,2 (c = 40)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............2 (c = 30)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)
................ (c = 0)
................0 (c = 10)"
`;

exports[`snapshot: 9 1`] = `
"1,8,0,6 (c = 190)
..0,1,6,8 (c = 190)
.... (c = 0)
....0,1 (c = 30)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......0,0 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........0 (c = 10)
.......... (c = 0)
....0,1,6 (c = 100)
...... (c = 0)
......0 (c = 10)
........ (c = 0)
......0,1 (c = 30)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........0,0 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........0 (c = 10)
............ (c = 0)
......1 (c = 20)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
......6 (c = 70)
........ (c = 0)
........0 (c = 10)
.......... (c = 0)
........3 (c = 40)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
........4 (c = 50)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..........3 (c = 40)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
............2 (c = 30)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)
................ (c = 0)
................0 (c = 10)
.................. (c = 0)
........5 (c = 60)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........2 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..........3 (c = 40)"
`;

exports[`snapshot: 10 1`] = `
"5,9,4,6,4,1,4,9,1,5 (c = 580)
..1,1,4,4,4,5,5,6,9,9 (c = 580)
.... (c = 0)
....1,1,4,4,4 (c = 190)
...... (c = 0)
......1,1 (c = 40)
........ (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........0,1 (c = 30)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........0,0 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............0 (c = 10)
.............. (c = 0)
........1,0 (c = 30)
..........0,1 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
............0,0 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............0 (c = 10)
................ (c = 0)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
.......... (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........0 (c = 10)
............ (c = 0)
..........0,0 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............0 (c = 10)
.............. (c = 0)
......1,1,4 (c = 90)
........ (c = 0)
........1 (c = 20)
.......... (c = 0)
..........0 (c = 10)
............ (c = 0)
........1,1 (c = 40)
.......... (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........1 (c = 20)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
..........0,1 (c = 30)
............ (c = 0)
............0 (c = 10)
.............. (c = 0)
............1 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
............0,0 (c = 20)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............0 (c = 10)
................ (c = 0)
..........1,0 (c = 30)
............0,1 (c = 30)
.............. (c = 0)
..............0 (c = 10)
................ (c = 0)
..............1 (c = 20)"
`;
